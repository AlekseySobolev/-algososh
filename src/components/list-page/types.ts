export interface ILinkedList<T> {
  append: (element: T) => void;
  prepend: (element: T) => void;
  deleteHead: () => void;
  deleteTail: () => void;
  addByIndex: (element: T, index: number) => void;
  deleteByIndex: (index: number) => void;
  toArray: () => (T | string)[];
  changeByIndex: (index: number, newValue: string) => void;
}

export enum ListOperation {
  Prepend = "prepend",
  Append = "append",
  DeleteHead = "deleteHead",
  DeleteTail = "deleteTail",
  AddByIndex = "addByIndex",
  DeleteByIndex = "deleteByIndex"
}