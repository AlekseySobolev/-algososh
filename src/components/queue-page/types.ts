export interface IQueue<T> {
  enqueue: (item: T) => void;
  dequeue: () => void;
  peak: () => T | null;
  last: () => T | null;
  getHead: ()  => number;
  getTail: ()  => number;
  clean: () => void;
  getContainer: () => (T | null)[] | null;
}

export enum QueueOperation {
  Queue = "queue",
  Dequeue = "dequeue",
  Clean = "clean"
}