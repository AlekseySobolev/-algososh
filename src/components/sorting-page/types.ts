export interface Step {
    currentArray: number[];
    sortedIndexes: number[];
    aIndex?: number;
    bIndex?: number;
}

export enum typeOfSort {
    Select = "select",
    Bubble = "bubble"
}

export enum sortDirection {
    Ascending = "ascending",
    Descending = "descending"
}
