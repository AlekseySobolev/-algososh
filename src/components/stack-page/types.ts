export interface IStack<T> {
    push: (item: T) => void;
    pop: () => void;
    clear: () => void;
  }

export enum StackOperation{
    Push = "push",
    Pop = "pop",
    Clear = "clear"
}