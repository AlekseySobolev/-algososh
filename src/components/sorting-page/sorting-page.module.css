.algoContainer{
    margin: auto;
    width: calc(100% - 410px * 2);
    
    display: flex;
    align-items: center;
    flex-direction: column;
    justify-content: space-between;
}

.btnBox{
    width: 100%;
    display: flex;
    align-items: center;
    margin-bottom: 48px;
}
.columnBox{
    display: flex;
    align-items: baseline; 
    gap: 10px;
    margin-bottom: 26px;
}
.radioBtn{
    display: flex;
    justify-content: space-between;
    min-width: 143px;
    margin-right: 52px;
    gap: 45px;
}
.ascBox{
    display: flex;
    gap: 12px;
    margin-right: 80px;
  }
  .btn{
    max-width: 205px;
  }