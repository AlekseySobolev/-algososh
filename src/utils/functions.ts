export const swap = (array: string[] | number[], firstIndex: number, secondIndex: number): string[] | number[] => {
    let temp;
    temp = array[firstIndex];
    array[firstIndex] = array[secondIndex];
    array[secondIndex] = temp;
    return array;
  };

  export const stringSort = (symbolArr: string[]): string[][] => {

    const arrayOfSortedArray: string[][] = [];
    arrayOfSortedArray.push([...symbolArr]);
    
    let i = 0;
    let j = symbolArr.length - 1;

    while(i < j){
      swap(symbolArr, i, j);
      arrayOfSortedArray.push([...symbolArr]);
      i++;
      j--; 
    }
    return arrayOfSortedArray;
  }