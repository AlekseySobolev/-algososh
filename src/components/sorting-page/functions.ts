import { Direction } from "../../types/direction";
import { ElementStates } from "../../types/element-states";
import { swap } from "../../utils/functions";
import { Step } from "./types";

const getIntFromInterval = (min: number, max: number) => {
    return Math.floor(Math.random() * (max - min + 1) + min)
}

export const getRandomArray = (min = 0, max = 100, minLength = 3, maxLength = 17) => {
    const result = new Array(getIntFromInterval(minLength, maxLength)).fill(0);
    return Array.from(new Set(result.map(() => getIntFromInterval(min, max))))
}

export const getColumnState = (index: number, isLastStep: boolean, currerntStep: Step): ElementStates => {
    if ([currerntStep.aIndex, currerntStep.bIndex].includes(index)) {
        return ElementStates.Changing;
    }

    if (currerntStep.sortedIndexes.includes(index) || isLastStep) {
        return ElementStates.Modified;
    }

    return ElementStates.Default;
}

export const bubbleSort = (arrForSort: number[], sortDirect: Direction): Step[] => {

    const sortSteps: Step[] = [];
    let itrNum = 0;
    let isSorted = false;

    do {

        isSorted = false;

        for (let i = 0; i < arrForSort.length - 1 - itrNum; i++) {
            if (sortDirect === Direction.Descending ? arrForSort[i] < arrForSort[i + 1] : arrForSort[i] > arrForSort[i + 1]) {
                swap(arrForSort, i, i + 1);
                isSorted = true;
            }
            sortSteps.push({ currentArray: [...arrForSort], sortedIndexes: [...(sortSteps[sortSteps.length - 1]?.sortedIndexes || [])], aIndex: i, bIndex: i + 1 });
        }

        ++itrNum;
        sortSteps[sortSteps.length - 1]?.sortedIndexes.push(arrForSort.length - itrNum);

    } while (isSorted)

    sortSteps.push({ currentArray: [...arrForSort], sortedIndexes: [...(sortSteps[sortSteps.length - 1]?.sortedIndexes || [])] });

    return sortSteps;
}

export const selectSort = (arrForSort: number[], sortDirect: Direction): Step[] => {

    const sortSteps: Step[] = [];
    let itrNum = 0;
    let isSorted = false;

    do {

        isSorted = false;

        for (let i = 0; i < arrForSort.length - 1 - itrNum; i++) {
            let maxInd = i;
            let j = i + 1;

            if (sortDirect === Direction.Descending ? arrForSort[maxInd] < arrForSort[j] : arrForSort[maxInd] > arrForSort[j]) {
                maxInd = j;
            }

            if (maxInd !== i) {
                swap(arrForSort, i, maxInd);
                isSorted = true;
            }
            sortSteps.push({ currentArray: [...arrForSort], sortedIndexes: [...(sortSteps[sortSteps.length - 1]?.sortedIndexes || [])], aIndex: itrNum, bIndex: itrNum + j});
        }

        ++itrNum;
        sortSteps[sortSteps.length - 1]?.sortedIndexes.push(arrForSort.length - itrNum);

    } while (isSorted)

    sortSteps.push({ currentArray: [...arrForSort], sortedIndexes: [...(sortSteps[sortSteps.length - 1]?.sortedIndexes || [])] });

    return sortSteps;

} 