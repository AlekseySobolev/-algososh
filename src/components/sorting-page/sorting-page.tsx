import React, { useRef, useState } from "react";
import { Button } from "../ui/button/button";
import { RadioInput } from "../ui/radio-input/radio-input";
import { SolutionLayout } from "../ui/solution-layout/solution-layout";
import { bubbleSort, getColumnState, getRandomArray, selectSort } from "./functions";
import { Step, typeOfSort } from "./types";
import styles from "./sorting.module.css";
import { Column } from "../ui/column/column";
import { Direction } from "../../types/direction";
import { SHORT_DELAY_IN_MS } from "../../constants/delays";

export const SortingPage: React.FC = () => {


  const randomArr = useRef<number[]>(getRandomArray());
  const timer = useRef<number>();
  const currentDirection = useRef<Direction>(Direction.Ascending);
  const [currentTypeOfSort, setCurrentTypeOfSort] = useState<typeOfSort>(typeOfSort.Select);
  const [currentSortStep, setCurrentSortStep] = useState(0);

  const [allSortSteps, setAllSortSteps] = useState<Step[]>([{
    currentArray: randomArr.current,
    sortedIndexes: []
  }]);

  const isAlgoInProcess = currentSortStep < allSortSteps.length - 1;


  const createNewArr = () => {
    randomArr.current = getRandomArray();
    setAllSortSteps([{
      currentArray: randomArr.current,
      sortedIndexes: []
    }]);
    setCurrentSortStep(0);
  }

  const makeSort = (currDirection: Direction, sortType: typeOfSort) => {
    let steps = [];
    if (sortType === typeOfSort.Bubble) {
      steps = bubbleSort(randomArr.current, currDirection);
    } else {
      steps = selectSort(randomArr.current, currDirection);
    }
    setAllSortSteps(steps);

    timer.current = window.setInterval(() => {

      if (steps.length > 0) {

        setCurrentSortStep((currentStep) => {
          const nextStep = currentStep + 1;

          if (nextStep > steps.length - 1 && timer.current) {
            window.clearInterval(timer.current);
            return currentStep;
          }

          return nextStep;
        });

      }

    }, SHORT_DELAY_IN_MS);
  }

  const ascClick = () => {

    setAllSortSteps([{
      currentArray: randomArr.current,
      sortedIndexes: []
    }]);
    setCurrentSortStep(0);

    currentDirection.current = Direction.Ascending;
    makeSort(currentDirection.current, currentTypeOfSort);
  }

  const descClick = () => {

    setAllSortSteps([{
      currentArray: randomArr.current,
      sortedIndexes: []
    }]);
    setCurrentSortStep(0);

    currentDirection.current = Direction.Descending;
    makeSort(currentDirection.current, currentTypeOfSort);
  }

  return (
    <SolutionLayout title="Сортировка массива">
      <div className={styles.algoContainer}>
        <div className={styles.btnBox}>
          <div className={styles.radioBtn}>
            <RadioInput
              label="Выбор"
              checked={currentTypeOfSort === typeOfSort.Select}
              onChange={() => setCurrentTypeOfSort(typeOfSort.Select)}
              disabled={isAlgoInProcess}
            />
            <RadioInput
              label="Пузырёк"
              checked={currentTypeOfSort === typeOfSort.Bubble}
              onChange={() => setCurrentTypeOfSort(typeOfSort.Bubble)}
              disabled={isAlgoInProcess}
            />
          </div>
          <div className={styles.ascBox}>
          <Button
            sorting={Direction.Ascending}
            text="По возрастанию"
            handleClick={ascClick}
            disabled={currentDirection.current !== Direction.Ascending && isAlgoInProcess}
            isLoader={currentDirection.current === Direction.Ascending && isAlgoInProcess}
            extraClass={styles.btn}
          />
          <Button
            sorting={Direction.Descending}
            text="По убыванию"
            handleClick={descClick}
            disabled={currentDirection.current !== Direction.Descending && isAlgoInProcess}
            isLoader={currentDirection.current === Direction.Descending && isAlgoInProcess}
            extraClass={styles.btn}
          />
          </div>
          
          <Button
            text="Новый массив"
            handleClick={createNewArr}
            disabled={isAlgoInProcess}
            extraClass={styles.btn}
          />
        </div>

        <div className={styles.columnBox}>
          {allSortSteps.length > 0 &&
            allSortSteps[currentSortStep].currentArray.map((currentNumber, index) => (
              <Column
                key={currentNumber}
                index={currentNumber}
                state={getColumnState(index, (currentSortStep > 0 && currentSortStep >= allSortSteps.length - 1), allSortSteps[currentSortStep])}
              />
            )

            )}
        </div>
      </div>
    </SolutionLayout>
  );
};
