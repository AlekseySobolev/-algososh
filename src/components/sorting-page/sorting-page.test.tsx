import { fireEvent, render, screen } from "@testing-library/react";
import '@testing-library/jest-dom/extend-expect';
import React from "react";
import { bubbleSort, selectSort } from '../../components/sorting-page/functions'
import { Direction } from '../../types/direction'

describe('sort algo tests',
    () => {

        it('should correct sort empty array', () => {

            let sortedArray = bubbleSort([], Direction.Ascending);
            let resultArray = [{ currentArray: [], sortedIndexes: [] }];
            expect(sortedArray).toEqual(resultArray);

            sortedArray = bubbleSort([], Direction.Descending);
            resultArray = [{ currentArray: [], sortedIndexes: [] }];
            expect(sortedArray).toEqual(resultArray);

            sortedArray = selectSort([], Direction.Ascending);
            resultArray = [{ currentArray: [], sortedIndexes: [] }];
            expect(sortedArray).toEqual(resultArray);

            sortedArray = selectSort([], Direction.Descending);
            resultArray = [{ currentArray: [], sortedIndexes: [] }];
            expect(sortedArray).toEqual(resultArray);

        })

        it('should correct sort array with single element', () => {
            
            let sortedArray = bubbleSort([1], Direction.Ascending);
            let resultArray = [{ currentArray: [1], sortedIndexes: [] }];
            expect(sortedArray).toEqual(resultArray);

            sortedArray = bubbleSort([1], Direction.Descending);
            resultArray = [{ currentArray: [1], sortedIndexes: [] }];
            expect(sortedArray).toEqual(resultArray);

            sortedArray = selectSort([1], Direction.Ascending);
            resultArray = [{ currentArray: [1], sortedIndexes: [] }];
            expect(sortedArray).toEqual(resultArray);

            sortedArray = selectSort([1], Direction.Descending);
            resultArray = [{ currentArray: [1], sortedIndexes: [] }];
            expect(sortedArray).toEqual(resultArray);

        })

        it('should correct sort array with many elements', () => {

            let sortedArray = bubbleSort([7, 5, 3], Direction.Ascending);
            let resultArray = { currentArray: [3, 5, 7], sortedIndexes: [2, 1, 0] };
            expect(sortedArray[sortedArray.length - 1]).toEqual(resultArray);

            sortedArray = bubbleSort([3, 5, 7], Direction.Descending);
            resultArray = { currentArray: [7, 5, 3], sortedIndexes: [2, 1, 0] };
            expect(sortedArray[sortedArray.length - 1]).toEqual(resultArray);
        })

    }
)