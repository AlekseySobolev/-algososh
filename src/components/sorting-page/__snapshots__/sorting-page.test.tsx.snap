// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sort algo tests should correct sort empty array 1`] = `
Array [
  Object {
    "currentArray": Array [],
    "sortedIndexes": Array [],
  },
]
`;
